src/community.c

```diff
@@ -3567,3 +3567,878 @@ int igraph_i_compare_communities_rand(const igraph_vector_t *v1,
+
+  /* Initialize neighboring cluster */
+  IGRAPH_CHECK(igraph_vector_bool_init(&neighbor_cluster_added, n));
+  IGRAPH_FINALLY(igraph_vector_bool_destroy, &neighbor_cluster_added);
+  IGRAPH_CHECK(igraph_vector_init(&neighbor_clusters, n));
+  IGRAPH_FINALLY(igraph_vector_destroy, &neighbor_clusters);
+
+  while (!igraph_dqueue_empty(&unstable_nodes))
+  {
+    long int v = (long int)igraph_dqueue_pop(&unstable_nodes);
+    long int best_cluster, current_cluster = VECTOR(*membership)[v];
+    long int degree, i;
+    igraph_vector_int_t *edges;
+
+    /* Remove node from current cluster */
+    VECTOR(cluster_weights)[current_cluster] -= VECTOR(*node_weights)[v];
+    VECTOR(nb_nodes_per_cluster)[current_cluster]--;
+    if (VECTOR(nb_nodes_per_cluster)[current_cluster] == 0)
+      igraph_stack_push(&empty_clusters, current_cluster);
+
+    /* Find out neighboring clusters */
+    c = (long int)igraph_stack_top(&empty_clusters);
+    VECTOR(neighbor_clusters)[0] = c;      
+    VECTOR(neighbor_cluster_added)[c] = 1;
+    nb_neigh_clusters = 1;
+
+    edges = igraph_inclist_get(edges_per_node, v);
+    degree = igraph_vector_int_size(edges);
+    for (i = 0; i < degree; i++)
+    {
+      long int e = VECTOR(*edges)[i];
+      long int u = (long int)IGRAPH_OTHER(graph, e, v);
+      c = VECTOR(*membership)[u];
+      if (!VECTOR(neighbor_cluster_added)[c])
+      {
+        VECTOR(neighbor_cluster_added)[c] = 1;
+        VECTOR(neighbor_clusters)[nb_neigh_clusters++] = c;      
+      }
+      VECTOR(edge_weights_per_cluster)[c] += VECTOR(*edge_weights)[e];
+    }
+
+    /* Calculate maximum diff */
+    best_cluster = current_cluster;
+    max_diff = VECTOR(edge_weights_per_cluster)[current_cluster] - VECTOR(*node_weights)[v] * VECTOR(cluster_weights)[current_cluster] * resolution_parameter;
+    for (i = 0; i < nb_neigh_clusters; i++)
+    {
+      c = VECTOR(neighbor_clusters)[i];
+      diff = VECTOR(edge_weights_per_cluster)[c] - VECTOR(*node_weights)[v] * VECTOR(cluster_weights)[c] * resolution_parameter;
+      if (diff > max_diff)
+      {
+        best_cluster = c;
+        max_diff = diff;
+      }
+      VECTOR(edge_weights_per_cluster)[c] = 0.0;
+      VECTOR(neighbor_cluster_added)[c] = 0;
+    }
+
+    /* Move node to best cluster */
+    VECTOR(cluster_weights)[best_cluster] += VECTOR(*node_weights)[v];
+    VECTOR(nb_nodes_per_cluster)[best_cluster]++;
+    if (best_cluster == igraph_stack_top(&empty_clusters))
+      igraph_stack_pop(&empty_clusters);
+
+    /* Mark node as stable */
+    VECTOR(node_is_stable)[v] = 1;
+
+    /* Add neighbours to unstable_nodes  */
+    if (best_cluster != current_cluster)
+    {
+        VECTOR(*membership)[v] = best_cluster;
+
+        for (i = 0; i < degree; i++)
+        {
+          long int e = VECTOR(*edges)[i];
+          long int u = (long int)IGRAPH_OTHER(graph, e, v);
+          if (!VECTOR(node_is_stable)[u] && VECTOR(*membership)[u] != best_cluster)
```

Comment:
Quite crucially, this should be `if(VECTOR(node_is_stable)[u] ...` (i.e. not the negation).